src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
                                 * genesis block rollback
                                 * etc -}
 
-data BlockEventRollback = BlockEventRollback
-    { _berInput    :: ![Blund SscGodTossing]
+instance IsBlockEventFailure BlockRollbackResult where
+    isBlockEventFailure = \case
+        BlockRollbackFailure -> True
+        _ -> False
+
+data BlockEventRollback' blund = BlockEventRollback
+    { _berInput    :: !(NewestFirst NE blund)
     , _berOutValid :: !BlockRollbackResult
+    } deriving (Functor, Foldable)
+
+makeLenses ''BlockEventRollback'
+
+instance IsBlockEventFailure (BlockEventRollback' blund) where
+    isBlockEventFailure = isBlockEventFailure . view berOutValid
+
+type BlockEventRollback = BlockEventRollback' BlundDefault
+
+data BlockEvent' blund
+    = BlkEvApply (BlockEventApply' blund)
+    | BlkEvRollback (BlockEventRollback' blund)
+    deriving (Functor, Foldable)
+
+instance IsBlockEventFailure (BlockEvent' blund) where
+    isBlockEventFailure = \case
+        BlkEvApply    a -> isBlockEventFailure a
+        BlkEvRollback a -> isBlockEventFailure a
+
+type BlockEvent = BlockEvent' BlundDefault
```

Comment:
Why did you parametrize everything with it instead of hardcoding `SscGodTossing`?